Description: enhance the cmake build scripts to:
 - allow targetting of multiple CUDA architectures (with -gencode=)
 - introduce DEFAULT_CUDA_ARCH to set the default CUDA arch target (-arch=)
Original Author: Jasper Grimm <jasper.grimm@york.ac.uk>
Updated by:
 - Ake Sandgren <ake.sandgren@hpc2n.umu.se>
     Modified to correctly set NVCC build flags and allow externally setting --std=xxx flag.
 - Jasper Grimm <jasper.grimm@york.ac.uk>
    Update for relion 5.0.0 beta

diff -Nru relion.orig/cmake/BuildTypes.cmake relion/cmake/BuildTypes.cmake
--- relion.orig/cmake/BuildTypes.cmake	2024-01-12 16:34:35.601247994 +0000
+++ relion/cmake/BuildTypes.cmake	2024-01-12 16:36:40.744752894 +0000
@@ -4,8 +4,8 @@
 # So -D__INTEL_COMPILER just manages to avoid compiling float128-targets (see fftw3.h, for instance).
 # Add -G to allow cuda-gdb to break inside kernels.
 if(CUDA)
-    set(EXTRA_NVCC_FLAGS "-D__INTEL_COMPILER --default-stream per-thread --std=c++14")
-    set(RELION_NVCC_FLAGS "${CUDARCH} ${WARN_DBL} ${EXTRA_NVCC_FLAGS}" CACHE STRING "" FORCE)
+    set(EXTRA_NVCC_FLAGS "-D__INTEL_COMPILER --default-stream per-thread")
+    set(RELION_NVCC_FLAGS "${WARN_DBL} ${EXTRA_NVCC_FLAGS}" CACHE STRING "" FORCE)
 elseif (HIP)
     if (${HIP_VERSION} VERSION_LESS "5.3" )
         set(EXTRA_HIPCC_FLAGS "-fgpu-default-stream=legacy -fno-gpu-rdc -munsafe-fp-atomics")
diff -Nru relion.orig/CMakeLists.txt relion/CMakeLists.txt
--- relion.orig/CMakeLists.txt	2024-01-12 16:34:35.601247994 +0000
+++ relion/CMakeLists.txt	2024-01-12 16:41:34.615595884 +0000
@@ -84,13 +84,20 @@
 
 set(CUDA_ARCH "" CACHE STRING "CUDA architecture to use")
 if (CUDA_ARCH STREQUAL "")
-	message(STATUS "Setting fallback CUDA_ARCH=50")
-	set(CUDARCH "-arch=sm_50")
+	message(STATUS "Setting fallback CUDA_ARCH=5.0")
+	set(CUDA_ARCH "5.0")
 else ()
 	message(STATUS "Using provided CUDA_ARCH=${CUDA_ARCH}")
-	set(CUDARCH "-arch=sm_${CUDA_ARCH}")
 endif ()
 
+if(NOT DEFINED DEFAULT_CUDA_ARCH)
+    string(REGEX MATCH "[0-9]+\\.[0-9]+" DEFAULT_CUDA_ARCH "${CUDA_ARCH}")
+    message(STATUS "Setting fallback DEFAULT_CUDA_ARCH=${DEFAULT_CUDA_ARCH}")
+endif()
+
+string(REPLACE "." "" DEFAULT_CUDA_SM ${DEFAULT_CUDA_ARCH})
+set(CUDA_SM "-arch=sm_${DEFAULT_CUDA_SM}")
+
 # -------------------------------------------------------------------FURTHER OPTIONS--
 
 # CUDA on by default, so check for other GPU/CPU-accelration request and possible conflicting dual-requests
@@ -204,6 +211,8 @@
       if(CUDA)
           add_definitions(-D_CUDA_ENABLED)
       endif()
+      cuda_select_nvcc_arch_flags(CUDA_GENCODE "${CUDA_ARCH}")
+      list(APPEND CUDA_NVCC_FLAGS ${CUDA_SM} ${CUDA_GENCODE})
     else(CUDA_FOUND)
         message(FATAL_ERROR "CUDA enabled but unlable to locate packages...")
     endif(CUDA_FOUND)
